modules/caddyevents/app.go

```diff
@@ -0,0 +1,318 @@
+
+	app.started = true
+
+	return nil
+}
+
+// Stop gracefully shuts down the app.
+func (app *App) Stop() error {
+	return nil
+}
+
+// Subscribe binds one or more event handlers to one or more events
+// according to the subscription s. For now, subscriptions can only
+// be created during the provision phase; new bindings cannot be
+// created after the events app has started.
+func (app *App) Subscribe(s *Subscription) error {
+	if app.started {
+		return fmt.Errorf("events already started; new subscriptions closed")
+	}
+
+	// handle special case of catch-alls (omission of event name or module space implies all)
+	if len(s.Events) == 0 {
+		s.Events = []string{""}
+	}
+	if len(s.Modules) == 0 {
+		s.Modules = []caddy.ModuleID{""}
+	}
+
+	for _, eventName := range s.Events {
+		if app.subscriptions[eventName] == nil {
+			app.subscriptions[eventName] = make(map[caddy.ModuleID][]Handler)
+		}
+		for _, originModule := range s.Modules {
+			app.subscriptions[eventName][originModule] = append(app.subscriptions[eventName][originModule], s.Handlers...)
+		}
+	}
+
+	return nil
+}
+
+// On is syntactic sugar for Subscribe() that binds a single handler
+// to a single event from any module. If the eventName is empty string,
+// it counts for all events.
+func (app *App) On(eventName string, handler Handler) error {
+	return app.Subscribe(&Subscription{
+		Events:   []string{eventName},
+		Handlers: []Handler{handler},
+	})
+}
+
+// Emit creates and dispatches an event named eventName to all relevant handlers with
+// the metadata data. Events are emitted and propagated synchronously. The returned Event
+// value will have any additional information from the invoked handlers.
+func (app *App) Emit(ctx caddy.Context, eventName string, data map[string]interface{}) Event {
+	id, err := uuid.NewRandom()
+	if err != nil {
+		app.logger.Error("failed generating new event ID", zap.Error(err))
+	}
+
+	eventName = strings.ToLower(eventName)
+
+	// TODO: make pointer?
+	e := Event{
+		id:     id,
+		ts:     time.Now(),
+		name:   eventName,
+		origin: ctx.Module(),
+		data:   data,
+	}
+
+	// add event info to replacer, make sure it's in the context
+	repl, ok := ctx.Context.Value(caddy.ReplacerCtxKey).(*caddy.Replacer)
+	if !ok {
+		repl = caddy.NewReplacer()
+		ctx.Context = context.WithValue(ctx.Context, caddy.ReplacerCtxKey, repl)
+	}
+	repl.Set("event", e)
+	repl.Set("event.id", e.id)
+	repl.Set("event.timestamp", e.ts)
+	repl.Set("event.name", eventName)
+	repl.Set("event.module", e.origin.CaddyModule().ID)
+	repl.Set("event.data", e.data)
+	repl.Map(func(key string) (interface{}, bool) {
+		if !strings.HasPrefix(key, "event.data.") {
+			return nil, false
+		}
+		key = strings.TrimPrefix(key, "event.data.")
+		if val, ok := data[key]; ok {
+			return val, true
+		}
+		return nil, false
+	})
+
+	// TODO: log that event was emitted... hm, maybe disable our logger by default? or only emit logs if any subscriptions are configured? or maybe debug level? hmm
```

Comment:
I suppose we should just `debug` log it for now and if it does get truly noisy we could do some hacking to make it off by default. But for now, we probably won't have _that_ many events (yet)

Logging the events that happen is super useful for users to discover what they can listen for. But obviously we should ideally document them too.

Reply:
Makes sense; done.